ipa-core/src/protocol/ipa_prf/aggregation/mod.rs

```diff
@@ -0,0 +1,257 @@
+    let bucket_ctx = ctx
+        .narrow(&Step::MoveToBucket)
+        .set_total_records(TotalRecords::Indeterminate);
+    // move each value to the correct bucket
+    let row_contributions_stream = process_stream_by_chunks(
+        contributions_stream,
+        AttributionOutputs {
+            attributed_breakdown_key_bits: vec![],
+            capped_attributed_trigger_value: vec![],
+        },
+        move |idx, chunk: AttributionOutputsChunk<N>| {
+            let record_id = RecordId::from(idx);
+            let ctx = bucket_ctx.clone();
+            async move {
+                bucket::move_single_value_to_bucket::<_, N>(
+                    ctx,
+                    record_id,
+                    chunk.attributed_breakdown_key_bits,
+                    chunk.capped_attributed_trigger_value,
+                    B,
+                    false,
+                )
+                .await
+            }
+        },
+        || AttributionOutputs {
+            attributed_breakdown_key_bits: Replicated::ZERO,
+            capped_attributed_trigger_value: Replicated::ZERO,
+        },
+    )
+    .then(|fut| async move { fut.await.map(Chunk::into_raw) });
+
+    let aggregation_input = Box::pin(
+        row_contributions_stream
+            .map_ok(|chunk| {
+                // Aggregation intermediate transpose
+                Vec::transposed_from(chunk.as_slice()).unwrap_infallible()
+            })
+            .try_flatten_iters::<BitDecomposed<_>, Vec<_>>(),
+    );
+
+    aggregate_values::<_, _, B>(ctx, aggregation_input, num_chunks * N).await
+}
+
+pub type AggResult<const B: usize> = Result<BitDecomposed<Replicated<Boolean, B>>, Error>;
+
+pub async fn aggregate_values<'fut, C, OV, const B: usize>(
+    ctx: C,
+    mut aggregated_stream: Pin<Box<dyn Stream<Item = AggResult<B>> + Send + 'fut>>,
+    mut num_rows: usize,
+) -> Result<Vec<Replicated<OV>>, Error>
+where
+    C: UpgradedContext<Boolean, Share = Replicated<Boolean>>,
+    OV: SharedValue + U128Conversions + CustomArray<Element = Boolean>,
+    Boolean: FieldSimd<B>,
+    Replicated<Boolean, B>: BooleanProtocols<C, Boolean, B>,
+    Vec<Replicated<OV>>:
+        for<'a> TransposeFrom<&'a BitDecomposed<Replicated<Boolean, B>>, Error = LengthError>,
+{
+    let mut depth = 0;
+    while num_rows > 1 {
+        // We reduce pairwise, passing through the odd record at the end if there is one, so the
+        // record count gets rounded down, but the number of outputs gets rounded up.
+        let par_agg_ctx = ctx
+            .narrow(&Step::Aggregate(depth))
+            .set_total_records(TotalRecords::Indeterminate);
+        num_rows = (num_rows + 1) / 2;
+        aggregated_stream = Box::pin(aggregated_stream.try_chunks(2).enumerate().then(
+            move |(i, chunk_res)| {
+                let ctx = par_agg_ctx.clone();
+                async move {
+                    match chunk_res {
+                        Err(e) => Err(e.1),
```

Comment:
[`try_chunks`](https://docs.rs/futures/latest/futures/stream/trait.TryStreamExt.html#method.try_chunks) combinator returns a tuple (buf, error) in case if the underlying stream failed. This code just ignores buffered items and returns the error which is the right thing to do imo

Reply:
I will add a comment.